Torque driver unable to get job status

Running Ert (cli) on the Azure cluster using the torque driver
Right after starting, a message like this is spit out for every job:```** Warning: failed to get job status for job:218 from file:/tmp/enkf-qstat-5956-365084```  
I do have write access to the `/tmp` folder, but the file is not created (regular users might have write access to that folder, but that's another issue).

The jobs actually are started and complete correctly, but Ert hangs, most likely because it's unable to detect that the jobs are completed

This seems to have reemerged.

```
qstat: Invalid credential
** Warning: failed to get job status for job:6725 from file:/tmp/enkf-qstat-25008-648985
Exception in thread Thread-6:
Traceback (most recent call last):
 File "/opt/rh/rh-python38/root/usr/lib64/python3.8/threading.py", line 932, in _bootstrap_inner
   self.run()
 File "/opt/rh/rh-python38/root/usr/lib64/python3.8/threading.py", line 870, in run
   self._target(*self._args, **self._kwargs)
 File "/prog/res/komodo/bleeding-py38-rhel7/root/lib64/python3.8/site-packages/res/job_queue/job_queue_node.py", line 218, in _job_monitor
   raise AssertionError(
AssertionError: Unexpected job status type after running job: JOB_QUEUE_STATUS_FAILURE
```

Shouldn't we try again if the error code here
https://github.com/equinor/ert/blob/d88cdd1c278c779609b514e7286bae57612d5a3e/libres/lib/job_queue/torque_driver.cpp#L527-L528
is non-zero?

--- a/src/libres/lib/job_queue/torque_driver.cpp
+++ b/src/libres/lib/job_queue/torque_driver.cpp
@@ -326,6 +326,7 @@ static void torque_debug(const torque_driver_type *driver, const char *fmt,
         }
         fprintf(driver->debug_stream, "\n");
         fsync(fileno(driver->debug_stream));
+        fflush(driver->debug_stream);
     }
 }
 
@@ -434,11 +435,27 @@ static int torque_driver_submit_shell_job(torque_driver_type *driver,
                 torque_debug(driver, "Submit arguments: %s",
                              stringlist_alloc_joined_string(remote_argv, " "));
                 char **argv = stringlist_alloc_char_ref(remote_argv);
-                int status = util_spawn_blocking(
-                    driver->qsub_cmd, stringlist_get_size(remote_argv),
-                    (const char **)argv, tmp_std_file, tmp_err_file);
-                if (status != 0) {
-                    torque_debug_spawn_status_info(driver, status);
+
+                /* The qsub command might fail intermittently for acceptable reasons,
+￼                  retry a couple of times with exponential sleep.  */
+                int return_value = -1;
+                int sleep_time = 2;
+                int max_sleep_time = 2 * 2 * 2 * 2; /* max 4 attempts */
+                while ((return_value != 0) & (sleep_time < max_sleep_time)) {
+                    return_value = util_spawn_blocking(
+                        driver->qsub_cmd, stringlist_get_size(remote_argv),
+                        (const char **)argv, tmp_std_file, tmp_err_file);
+                    if (return_value != 0) {
+                        torque_debug(
+                            driver,
+                            "qsub failed for job %s, retrying in %d seconds",
+                            job_name, sleep_time);
+                        sleep(sleep_time);
+                        sleep_time *= 2;
+                    }
+                }
+                if (return_value != 0) {
+                    torque_debug_spawn_status_info(driver, return_value);
                 }
                 free(argv);
                 stringlist_free(remote_argv);
@@ -514,26 +531,49 @@ void *torque_driver_submit_job(void *__driver, const char *submit_cmd,
 static job_status_type
 torque_driver_get_qstat_status(torque_driver_type *driver,
                                const char *jobnr_char) {
-    char *tmp_file = (char *)util_alloc_tmp_file("/tmp", "enkf-qstat", true);
+    char *tmp_std_file =
+        (char *)util_alloc_tmp_file("/tmp", "ert-qstat-std", true);
+    char *tmp_err_file =
+        (char *)util_alloc_tmp_file("/tmp", "ert-qstat-err", true);
     job_status_type status = JOB_QUEUE_STATUS_FAILURE;
 
     {
         const char **argv = (const char **)util_calloc(1, sizeof *argv);
         argv[0] = jobnr_char;
 
-        util_spawn_blocking(driver->qstat_cmd, 1, (const char **)argv, tmp_file,
-                            NULL);
+        /* The qstat command might fail intermittently for acceptable reasons,
+           retry a couple of times with exponential sleep. ERT pings qstat
+           every second for every realization, thus the initial sleep time
+           is 2 seconds. */
+        int return_value = -1;
+        int sleep_time = 2;                 /* seconds */
+        int max_sleep_time = 2 * 2 * 2 * 2; /* max 4 attempts */
+        while ((return_value != 0) & (sleep_time < max_sleep_time)) {
+            return_value =
+                util_spawn_blocking(driver->qstat_cmd, 1, (const char **)argv,
+                                    tmp_std_file, tmp_err_file);
+            if (return_value != 0) {
+                torque_debug(driver,
+                             "qstat failed for job %s, retrying in %d seconds",
+                             jobnr_char, sleep_time);
+                sleep(sleep_time);
+                sleep_time *= 2;
+            }
+        }
         free(argv);
     }
 
-    if (fs::exists(tmp_file)) {
-        status = torque_driver_parse_status(tmp_file, jobnr_char);
-        unlink(tmp_file);
+    if (fs::exists(tmp_std_file)) {
+        status = torque_driver_parse_status(tmp_std_file, jobnr_char);
+        unlink(tmp_std_file);
+        unlink(tmp_err_file);
     } else
-        fprintf(stderr, "No such file: %s - reading qstat status failed \n",
-                tmp_file);
-
-    free(tmp_file);
+        fprintf(stderr,
+                "No such file: %s - reading qstat status failed\n"
+                "stderr: %s\n",
+                tmp_std_file, tmp_err_file);
+    free(tmp_std_file);
+    free(tmp_err_file);
 
     return status;
 }

--- a/tests/libres_tests/res/job_queue/test_job_queue_manager.py
+++ b/tests/libres_tests/res/job_queue/test_job_queue_manager.py
@@ -1,9 +1,9 @@
 import os
 import stat
+from dataclasses import dataclass
 from pathlib import Path
 from threading import BoundedSemaphore
-from typing import List
-from dataclasses import dataclass
+from typing import Callable, List, TypedDict
 
 import pytest
 
@@ -23,6 +23,15 @@ class RunArg:
     iens: int
 
 
+class Config(TypedDict):
+    job_script: str
+    num_cpu: int
+    job_name: str
+    run_path: str
+    ok_callback: Callable
+    exit_callback: Callable
+
+
 def dummy_ok_callback(args):
     print(f"success {args[1]}")
     (Path(args[1]) / "OK").write_text("success", encoding="utf-8")
@@ -34,7 +43,7 @@ def dummy_exit_callback(args):
     Path("ERROR").write_text("failure", encoding="utf-8")
 
 
-DUMMY_CONFIG = {
+DUMMY_CONFIG: Config = {
     "job_script": "job_script.py",
     "num_cpu": 1,
     "job_name": "dummy_job_{}",
@@ -43,11 +52,12 @@ DUMMY_CONFIG = {
     "exit_callback": dummy_exit_callback,
 }
 
-SIMPLE_SCRIPT = """#!/usr/bin/env python
-with open('STATUS', 'w') as f:
-   f.write('finished successfully')
+SIMPLE_SCRIPT = """#!/bin/sh
+echo "finished successfully" > STATUS
 """
 
+# This script is susceptible to race conditions. Python works
+# better than sh.
 FAILING_SCRIPT = """#!/usr/bin/env python
 import sys
 with open("one_byte_pr_invocation", "a") as f:
@@ -55,10 +65,8 @@ with open("one_byte_pr_invocation", "a") as f:
 sys.exit(1)
 """
 
-MOCK_BSUB = """#!/usr/bin/env python
-import sys
-with open("test.out", "w") as filehandle:
-    filehandle.write(" ".join(sys.argv))
+MOCK_BSUB = """#!/bin/sh
+echo "$@" > test.out
 """
 """A dummy bsub script that instead of submitting a job to an LSF cluster
 writes the arguments it got to a file called test.out, mimicking what
@@ -106,20 +114,21 @@ def test_num_cpu_submitted_correctly_lsf(tmpdir):
 
     script = Path(DUMMY_CONFIG["job_script"])
     script.write_text(SIMPLE_SCRIPT, encoding="utf-8")
-    script.chmod(stat.S_IRWXU | stat.S_IRWXO | stat.S_IRWXG)
+    script.chmod(stat.S_IRWXU)
 
     bsub = Path("bsub")
     bsub.write_text(MOCK_BSUB, encoding="utf-8")
-    bsub.chmod(stat.S_IRWXU | stat.S_IRWXO | stat.S_IRWXG)
+    bsub.chmod(stat.S_IRWXU)
 
     job_id = 0
     num_cpus = 4
     os.mkdir(DUMMY_CONFIG["run_path"].format(job_id))
+
     job = JobQueueNode(
         job_script=DUMMY_CONFIG["job_script"],
         job_name=DUMMY_CONFIG["job_name"].format(job_id),
         run_path=os.path.realpath(DUMMY_CONFIG["run_path"].format(job_id)),
-        num_cpu=num_cpus,
+        num_cpu=4,
         status_file="STATUS",
         ok_file="OK",
         exit_file="ERROR",

--- a/tests/libres_tests/res/job_queue/test_job_queue_manager_torque.py
+++ b/tests/libres_tests/res/job_queue/test_job_queue_manager_torque.py
@@ -1,0 +1,181 @@
+import os
+import stat
+from dataclasses import dataclass
+from pathlib import Path
+from threading import BoundedSemaphore
+from typing import Callable, TypedDict
+
+import pytest
+
+from res._lib.model_callbacks import LoadStatus
+from res.job_queue import Driver, JobQueueNode, QueueDriverEnum
+
+
+@dataclass
+class RunArg:
+    iens: int
+
+
+class JobConfig(TypedDict):
+    job_script: str
+    num_cpu: int
+    job_name: str
+    run_path: str
+    ok_callback: Callable
+    exit_callback: Callable
+
+
+def dummy_ok_callback(args):
+    (Path(args[1]) / "OK").write_text("success", encoding="utf-8")
+    return (LoadStatus.LOAD_SUCCESSFUL, "")
+
+
+def dummy_exit_callback(args):
+    Path("ERROR").write_text("failure", encoding="utf-8")
+
+
+DUMMY_CONFIG: JobConfig = {
+    "job_script": "job_script.py",
+    "num_cpu": 1,
+    "job_name": "dummy_job_{}",
+    "run_path": "dummy_path_{}",
+    "ok_callback": dummy_ok_callback,
+    "exit_callback": dummy_exit_callback,
+}
+
+SIMPLE_SCRIPT = """#!/bin/sh
+echo "finished successfully" > STATUS
+"""
+
+# This script is susceptible to race conditions. Python works
+# better than sh.
+FAILING_SCRIPT = """#!/usr/bin/env python
+import sys
+with open("one_byte_pr_invocation", "a") as f:
+    f.write(".")
+sys.exit(1)
+"""
+
+MOCK_QSUB = """#!/bin/sh
+echo "torque job submitted" > job_output
+echo "$@" >> job_output
+echo "10001.s034-lcam"
+exit 0
+"""
+
+# A qsub shell script that will fail on the first invocation, but succeed on the
+# second (by persisting its state in the current working directory)
+FLAKY_QSUB = """#!/bin/sh
+if [ -s firstwasflaky ]; then
+    echo ok > job_output
+    echo "10001.s034-lcam"
+    exit 0
+fi
+echo "it was" > firstwasflaky
+exit 1
+"""
+
+
+def create_qstat_output(
+    state: str, job_id: str = "10001.s034-lcam", bash=False, bashindent=""
+):
+    assert len(state) == 1
+    mocked_output = f"""Job id            Name             User              Time Use S Queue
+----------------  ---------------- ----------------  -------- - -----
+{job_id: <16}  MyMockedJob      rms                      0 {state:<1} normal   100
+"""  # noqa
+    if bash:
+        return "\n".join(
+            [f'{bashindent}echo "{line}"' for line in mocked_output.splitlines()]
+        )
+    return mocked_output
+
+
+# A qstat script that works as expected:
+MOCK_QSTAT = "#!/bin/sh\n" + create_qstat_output(state="E", bash=True)
+
+# A qstat shell script that will fail on the first invocation, but succeed on
+# the second (by persisting its state in the current working directory)
+FLAKY_QSTAT = (
+    """#!/bin/sh
+sleep 1
+if [ -s firstwasflaky ]; then
+"""
+    + create_qstat_output(state="E", bash=True, bashindent="    ")
+    + """
+    exit 0
+fi
+echo "it was" > firstwasflaky
+# These stderr messages should be swallowed and muted by driver:
+if [ $RANDOM -le 10000 ]; then
+    echo "qstat: Invalid credential 10001.s034-lcam" >&2
+else
+    echo "qstat: Invalid credential" >&2
+fi
+exit 1
+"""
+)
+
+
+@pytest.mark.parametrize(
+    "qsub_script, qstat_script",
+    [
+        pytest.param(MOCK_QSUB, MOCK_QSTAT, id="none_flaky"),
+        pytest.param(FLAKY_QSUB, MOCK_QSTAT, id="flaky_qsub"),
+        pytest.param(MOCK_QSUB, FLAKY_QSTAT, id="flaky_qstat"),
+        pytest.param(FLAKY_QSUB, FLAKY_QSTAT, id="all_flaky"),
+    ],
+)
+def test_run_torque_job(tmpdir, qsub_script, qstat_script):
+    """Verify that the torque driver will succeed in submitting and
+    monitoring torque jobs even when the Torque commands qsub and qstat
+    are flaky.
+
+    A flaky torque command is a shell script that sometimes but not
+    always returns with a non-zero exit code."""
+    os.chdir(tmpdir)
+    os.putenv("PATH", os.getcwd() + ":" + os.getenv("PATH"))
+    driver = Driver(driver_type=QueueDriverEnum.TORQUE_DRIVER, max_running=1)
+
+    script = Path(DUMMY_CONFIG["job_script"])
+    script.write_text(SIMPLE_SCRIPT, encoding="utf-8")
+    script.chmod(stat.S_IRWXU)
+
+    qsub = Path("qsub")
+    qsub.write_text(qsub_script, encoding="utf-8")
+    qsub.chmod(stat.S_IRWXU)
+
+    qstat = Path("qstat")
+    qstat.write_text(qstat_script, encoding="utf-8")
+    qstat.chmod(stat.S_IRWXU)
+
+    job_id = 0
+    runpath = Path(DUMMY_CONFIG["run_path"].format(job_id))
+    runpath.mkdir()
+
+    job = JobQueueNode(
+        job_script=DUMMY_CONFIG["job_script"],
+        job_name=DUMMY_CONFIG["job_name"].format(job_id),
+        run_path=os.path.realpath(DUMMY_CONFIG["run_path"].format(job_id)),
+        num_cpu=1,
+        status_file="STATUS",
+        ok_file="OK",
+        exit_file="ERROR",
+        done_callback_function=DUMMY_CONFIG["ok_callback"],
+        exit_callback_function=DUMMY_CONFIG["exit_callback"],
+        callback_arguments=[
+            RunArg(iens=job_id),
+            Path(DUMMY_CONFIG["run_path"].format(job_id)).resolve(),
+        ],
+    )
+
+    pool_sema = BoundedSemaphore(value=2)
+    job.run(driver, pool_sema)
+    job.wait_for()
+
+    # This file is supposed created by the job that the qsub script points to,
+    # but here it is created by the mocked qsub.
+    assert Path("job_output").exists()
+
+    # The "done" callback:
+    assert (runpath / "OK").read_text(encoding="utf-8") == "success"

